src/x86_64/interrupt.c

```diff
@@ -330,65 +267,23 @@ void common_handler()
     }
 }
 
-heap interrupt_vectors;
+static heap interrupt_vector_heap;
 
-// actually allocate the virtual  - put in the tree
-static void enable_lapic(heap pages)
+u64 allocate_interrupt(void)
 {
-    // there is an msr that moves the physical
-    u64 lapic = 0xfee00000;
-    
-    map(u64_from_pointer(apic_base), lapic, PAGESIZE, PAGE_DEV_FLAGS, pages);
-    
-    // turn on the svr, then enable three lines
-    apic_write(APIC_SPURIOUS, *(unsigned int *)(apic_base + APIC_SPURIOUS) | APIC_SW_ENABLE);
-    apic_write(APIC_LVT_LINT0, APIC_DISABLE);
-    apic_write(APIC_LVT_LINT1, APIC_DISABLE);
-    apic_write(APIC_LVT_ERR, allocate_u64(interrupt_vectors, 1));
+    return allocate_u64(interrupt_vector_heap, 1);
 }
 
-
-void register_interrupt(int vector, thunk t)
+void deallocate_interrupt(u64 irq)
 {
-    handlers[vector] = t;
+    deallocate_u64(interrupt_vector_heap, irq, 1);
 }
 
-static u32 apic_timer_cal_sec;
-
-/* ugh, don't want to slow down the boot this much...see if we can
-   trim this some more */
-#define CALIBRATE_DURATION_MS 10
-void calibrate_lapic_timer()
-{
-    apic_write(APIC_TMRINITCNT, -1u);
-    kern_sleep(milliseconds(10));
-    u32 delta = -1u - apic_read(APIC_TMRCURRCNT);
-    apic_set(APIC_LVT_TMR, APIC_LVT_INTMASK);
-    apic_timer_cal_sec = (1000 / CALIBRATE_DURATION_MS) * delta;
-}
-
-void lapic_runloop_timer(timestamp interval)
-{
-    /* interval * apic_timer_cal_sec / second */
-    u32 cnt = (((u128)interval) * apic_timer_cal_sec) >> 32;
-    apic_clear(APIC_LVT_TMR, APIC_LVT_INTMASK);
-    apic_write(APIC_TMRINITCNT, cnt);
-}
-
-static CLOSURE_0_0(int_ignore, void);
-static void int_ignore(void) {}
-
-void configure_lapic_timer(heap h)
+void register_interrupt(int vector, thunk t)
 {
-    apic_write(APIC_TMRDIV, 3 /* 16 */);
-    int v = allocate_u64(interrupt_vectors, 1);
-    apic_write(APIC_LVT_TMR, v); /* one shot */
-    register_interrupt(v, closure(h, int_ignore));
-    calibrate_lapic_timer();
+    handlers[vector] = t;
```

Comment:
may be helpful to have a debug check for erroneously overwriting assigned vectors ..

Reply:
Fixed.